File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
        # res = Optional[ListNode]
        # if list1 == None:
        #     return list2
        # elif list2 == None:
        #     return list1
        # else:
        #     if list1.val < list2.val:
        #         list1.next = self.mergeTwoLists(list1.next, list2)
        #         return list1
        #     else:
        #         list2.next = self.mergeTwoLists(list1, list2.next)
        #         return list2
        head = res = ListNode()

        while list1 and list2:
            if list1.val < list2.val:
                head.next = list1
                list1 = list1.next
            else:
                head.next = list2
                list2 = list2.next
            head = head.next

        # update head to list1 not null or list2 not null
        head.next = list1 or list2
        return res.next
           

File: Data Structures & Algorithms/reorder-linked-list/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    void reorderList(ListNode* head) {
        // if(head == nullptr){
        //     return;
        // }
        // vector<ListNode*> nodeList;
        // ListNode* curr = head;
        // while(curr){
        //     nodeList.push_back(curr);
        //     curr = curr->next;
        // }
        // ListNode* res = head;
        // for(int i = 0; i < (nodeList.size()+1)/2; i++){
        //     res->next = nodeList[i];
        //     res = res->next;
        //     res->next = nodeList[nodeList.size() - 1  - i];
        //     res = res->next;
        // }
        // // make the last element in linkedlist nullptr 
        // res->next = nullptr;
        // int i = 0;
        // int j = nodeList.size()-1;
        // while (i < j){
        //     nodeList[i++]->next = nodeList[j];
        //     if(i >= j){
        //         break;
        //     }
        //     nodeList[j--]->next = nodeList[i];
        // }
        // nodeList[i]->next = nullptr;

        
        // use slow pointer and fast pointer to get the middle
        ListNode* slow = head;
        ListNode* fast = head;

        while(fast && fast->next){
            fast = fast->next->next;
            slow = slow->next;
        }
        // reverse the second part
        ListNode* second = slow->next;
        slow->next = nullptr;
        ListNode* curr = second;
        ListNode* prev = nullptr;

        while(curr){
            ListNode* temp = curr->next;
            curr->next = prev;
            prev = curr;
            curr = temp;
        }
        second = prev;
        ListNode* first = head;
        // merge to one list 
        while(second){
            ListNode* temp1 = first->next;
            ListNode* temp2 = second->next;
            first->next = second;
            second->next = temp1;
            first = temp1;
            second = temp2;
        }
        // reverse the 
    }
};
